### PR TITLE
actionify index policy action

### PR DESCRIPTION
**Pull request by bowenlan-amzn**
*Issue #, if available:*
#292 

*Description of changes:*
actionify index policy API, refactor our API code so we have an action name to each API

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by qreshi — DISMISSED** [review]

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/action/IndexPriorityActionConfig.kt [hunk 1]
```diff
@@ -54,6 +56,19 @@ data class IndexPriorityActionConfig(
...
 
+    @Throws(IOException::class)
+    constructor(sin: StreamInput) : this(
+        indexPriority = sin.readInt(),
+        index = sin.readInt()
+    )
+
+    @Throws(IOException::class)
+    override fun writeTo(out: StreamOutput) {
+        super.writeTo(out)
```
Comment (dbbaughe) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/action/IndexPriorityActionConfig.kt:67]: Here too, seems like all action configs in this PR have issue.

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/action/SnapshotActionConfig.kt [hunk 2]
```diff
@@ -57,6 +59,21 @@ data class SnapshotActionConfig(
         managedIndexMetaData: ManagedIndexMetaData
     ): Action = SnapshotAction(clusterService, client, managedIndexMetaData, this)
 
+    @Throws(IOException::class)
+    constructor(sin: StreamInput) : this(
+        repository = sin.readOptionalString(),
```
Comment (dbbaughe) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/action/SnapshotActionConfig.kt:64]: Confused why repo and snapshot are nullable in the primary constructor and then we have an init that requires them not to be null, perhaps it was just missed in a previous review. Can you change these to not be nullable so you don't have to do read optional?

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/action/AllocationActionConfig.kt [hunk 3]
```diff
@@ -61,12 +63,36 @@ data class AllocationActionConfig(
...
+    )
+
+    @Throws(IOException::class)
+    override fun writeTo(out: StreamOutput) {
+        super.writeTo(out)
+        out.writeMap(require)
+        out.writeMap(include)
+        out.writeMap(exclude)
+        out.writeBoolean(waitFor)
+        out.writeInt(index)
```
Comment (dbbaughe) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/action/AllocationActionConfig.kt:82]: I believe the action index is already being written in the ActionConfig class which you're calling with super, so shouldn't need to write and read on all subclassed action configs.

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/destination/Chime.kt [hunk 4]
```diff
@@ -45,6 +48,14 @@ data class Chime(val url: String?) : ToXContent {
                 .endObject()
     }
 
+    constructor(sin: StreamInput) : this(
```
Comment (dbbaughe) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/destination/Chime.kt:51]: Similar to other place -- not sure why url is nullable when we require it not ot be null in the init, can we just change it to not nullable and not have this optional?

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/destination/CustomWebhook.kt [hunk 5]
```diff
@@ -62,6 +65,30 @@ data class CustomWebhook(
                 .endObject()
     }
 
+    constructor(sin: StreamInput) : this(
+        sin.readOptionalString(),
+        sin.readOptionalString(),
+        sin.readOptionalString(),
+        sin.readInt(),
+        sin.readOptionalString(),
+        suppressWarning(sin.readMap()),
```
Comment (dbbaughe) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/destination/CustomWebhook.kt:74]: Same question about usage of suppressWarning here

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/destination/Slack.kt [hunk 6]
```diff
@@ -33,7 +36,7 @@ import java.lang.IllegalStateException
  * Temporary import from alerting, this will be removed once we pull notifications out of
  * alerting so all plugins can consume and use.
  */
-data class Slack(val url: String?) : ToXContent {
+data class Slack(val url: String?) : ToXContent, Writeable {
```
Comment (dbbaughe) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/destination/Slack.kt:39]: Same as chime, can we just change this to not nullable?

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/transport/action/indexpolicy/IndexPolicyRequest.kt [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        this.policy = policy
+        this.seqNo = seqNo
+        this.primaryTerm = primaryTerm
+        this.refreshPolicy = refreshPolicy
+    }
+
+    @Throws(IOException::class)
+    constructor(sin: StreamInput) : this(
+        policyID = sin.readString(),
+        policy = sin.readOptionalWriteable(::Policy) as Policy,
```
Comment (dbbaughe) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/transport/action/indexpolicy/IndexPolicyRequest.kt:52]: Not optional

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/transport/action/indexpolicy/IndexPolicyResponse.kt [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+        this.status = status
+    }
+
+    @Throws(IOException::class)
+    constructor(sin: StreamInput) : this(
+        id = sin.readString(),
+        version = sin.readLong(),
+        primaryTerm = sin.readLong(),
+        seqNo = sin.readLong(),
+        policy = sin.readOptionalWriteable(::Policy) as Policy,
```
Comment (dbbaughe) [inline src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/transport/action/indexpolicy/IndexPolicyResponse.kt:63]: Not optional

File: src/test/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/transport/action/indexpolicy/IndexPolicyResponseTests.kt [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+import org.elasticsearch.common.io.stream.BytesStreamOutput
+import org.elasticsearch.common.io.stream.StreamInput
+import org.elasticsearch.rest.RestStatus
+import org.elasticsearch.test.ESTestCase
+import java.time.Instant
+import java.time.temporal.ChronoUnit
+
+class IndexPolicyResponseTests : ESTestCase() {
+
+    fun `test index policy response index priority action`() {
```
Comment (dbbaughe) [inline src/test/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/transport/action/indexpolicy/IndexPolicyResponseTests.kt:31]: What is the difference between this test and the other index priority test above?

File: src/test/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/ActionConfigTests.kt [hunk 10]
```diff
@@ -75,4 +83,42 @@ class ActionConfigTests : ESTestCase() {
             randomAllocationActionConfig()
         }
     }
+
```
Comment (dbbaughe) [inline src/test/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/ActionConfigTests.kt:86]: Add some tests for an action config combined with retry/timeout settings.

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/Transition.kt [hunk 11]
```diff
@@ -96,6 +109,18 @@ data class Conditions(
         return builder.endObject()
     }
 
+    constructor(sin: StreamInput) : this(
```
[the remark this answers is not in the document]
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/Transition.kt:112]: awkward, last time I am doing this, Schedule hasn't been changed to Writeable and I forgot to call out, fortunately it's released in job scheduler 1.10.1.1 now [link](https://github.com/opendistro-for-elasticsearch/job-scheduler/commit/28f0449e98fa1fc1fee1061babe48b4e4c550da0)

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/action/AllocationActionConfig.kt [hunk 12]
```diff
@@ -61,12 +63,36 @@ data class AllocationActionConfig(
             .endObject()
     }
 
+    @Throws(IOException::class)
+    constructor(sin: StreamInput) : this(
```
[the remark this answers is not in the document]
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/action/AllocationActionConfig.kt:67]: sin.readMap() return Map<String, Object>, so either suppress it or cast to Map<String, String>

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/action/AllocationActionConfig.kt [hunk 13]
```diff
@@ -61,12 +63,36 @@ data class AllocationActionConfig(
...
+        require = suppressWarning(sin.readMap()),
+        include = suppressWarning(sin.readMap()),
+        exclude = suppressWarning(sin.readMap()),
+        waitFor = sin.readBoolean(),
+        index = sin.readInt()
+    )
+
+    @Throws(IOException::class)
+    override fun writeTo(out: StreamOutput) {
+        super.writeTo(out)
```
[the remark this answers is not in the document]
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/action/AllocationActionConfig.kt:77]: I am struggled here, my thought is:
1. all these specific actionConfig is abstract to ActionConfig
2. for streamInput, we use ActionConfig.fromStreamInput for all specific actionConfig, just like parse in ActionConfig
3. for writeTo, each specific actionConfig override and has it's own writeTo, just like toXContent

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/action/SnapshotActionConfig.kt (hunk 2 above)
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/action/SnapshotActionConfig.kt:64]: good point, so I just remove optional then, will leave this `String? and require(repository != null)` combination, looks informative haha

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/action/AllocationActionConfig.kt (hunk 3 above)
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/action/AllocationActionConfig.kt:82]: yes, it's duplicate, but it's needed in ActionConfig fromStreamInput

File: src/test/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/transport/action/indexpolicy/IndexPolicyResponseTests.kt (hunk 9 above)
Comment (bowenlan-amzn) [reply src/test/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/transport/action/indexpolicy/IndexPolicyResponseTests.kt:31]: 😀 no difference, first want to test ActionConfig, then write then separately in model/ActionConfig tests

**Review by dbbaughe — APPROVED** [review]

File: src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/action/SnapshotActionConfig.kt (hunk 2 above)
Comment (bowenlan-amzn) [reply src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/model/action/SnapshotActionConfig.kt]: will do require in parse

**Review by dbbaughe — APPROVED** [review]

**Review by ftianli-amzn — APPROVED** [review]

**Pull request merged**